src/qutip_qip/decompose/decompose_two_qubit_gate.py

```diff
@@ -0,0 +1,293 @@
+        single_qubit_U, N=2, control=1, target=0, control_value=1
+    )
+    return two_level_U5
+
+
+def _decompose_U6(input_array):
+    """Decomposes a two-level unitary into ABC and CNOT when target =1 and
+    the control value is 1.
+    Based on Lemma 5.1 of arXiv:quant-ph/9503016v1
+    """
+    U = np.array(
+        [[input_array[2][2], input_array[2][3]],
+            [input_array[3][2], input_array[3][3]]]
+    )
+    check_gate(Qobj(U), num_qubits=1)
+    gate_list = decompose_one_qubit_gate(Qobj(U), "ZYZ_PauliX")
+
+    # Change the gate targets for ABC and X gates from default = 0
+    # First 2 indices are for A, next 2 are for B and last is for C
+    # Note that we have avoided changing the Pauli X indices because it is
+    # easier to define a gate with the target qubit.
+    target_indices_to_change = [0, 1, 3, 4, 6]
+    for i in target_indices_to_change:
+        gate_list[i].targets = [1]
+
+    # last is phase whose target is not changed based on Lemma 5.2 of
+    # arXiv:quant-ph/9503016v1 but the gate type is changed
+    phase_angle = gate_list[-1].arg_value
+    phase_gate = Gate(
+        "PHASEGATE",
+        targets=[0],
+        arg_value=phase_angle,
+        arg_label=r"{:0.2f} \times \pi".format(phase_angle / np.pi),
+    )
+    Pauli_X = Gate("X", targets=[1], classical_controls=[0])
+
+    CNOT_ctrl_0 = Gate("CNOT", controls=0, targets=1)
+    return (
+        gate_list[0],
+        gate_list[1],
+        Pauli_X,
+        CNOT_ctrl_0,
+        Pauli_X,
+        gate_list[3],
+        gate_list[4],
+        Pauli_X,
+        CNOT_ctrl_0,
+        Pauli_X,
+        gate_list[6],
+        phase_gate,
+    )
+
+
+def _decompose_U5(input_array):
+    """Decomposes a two-level unitary into ABC and CNOT when target = 0 and
+    control value is 1.
+    Based on Lemma 5.1 of arXiv:quant-ph/9503016v1 with flipped targets.
+    """
+    U = np.array(
+        [[input_array[1][1], input_array[1][3]],
+            [input_array[3][1], input_array[3][3]]]
+    )
+    check_gate(Qobj(U), num_qubits=1)
+    gate_list = decompose_one_qubit_gate(Qobj(U), "ZYZ_PauliX")
+
+    # No need to change targets of A, B, C - default is already 0
+    # last is phase whose target is changed based on Lemma 5.2 of
+    # arXiv:quant-ph/9503016v1 alongwith the gate type
+    phase_angle = gate_list[-1].arg_value
+    phase_gate = Gate(
+        "PHASEGATE",
+        targets=[1],
+        arg_value=phase_angle,
+        arg_label=r"{:0.2f} \times \pi".format(phase_angle / np.pi),
+    )
+    Pauli_X = Gate("X", targets=[0], classical_controls=[0])
+
+    CNOT_ctrl_1 = Gate("CNOT", controls=1, targets=0)
+    return (
+        gate_list[0],
+        gate_list[1],
+        Pauli_X,
+        CNOT_ctrl_1,
+        Pauli_X,
+        gate_list[3],
+        gate_list[4],
+        Pauli_X,
+        CNOT_ctrl_1,
+        Pauli_X,
+        gate_list[6],
+        phase_gate,
+    )
+
+
+def _decompose_to_two_level_arrays(input_gate):
```

Comment:
This function will be discarded after a general function decomposing a n qubit gate to two level arrays is defined. 

The plan is to iterate over the row and column indices in a for loop to avoid manually [defining indices for the decomposition. ](https://github.com/qutip/qutip-qip/pull/70/files#diff-4fd62396e554b500c5f57c03c2e6cf0123b85fb4ecb549a57b97f8d120a25f62R141)